File: main/src/main/scala/org/clulab/dynet/GreedyForwardLayer.scala
package org.clulab.dynet

import java.io.PrintWriter
import edu.cmu.dynet.{Dim, Expression, ExpressionVector, LookupParameter, Parameter, ParameterCollection}
import org.clulab.dynet.Utils.{ByLineFloatBuilder, ByLineIntBuilder, ByLineStringBuilder, ByLineStringMapBuilder, fromIndexToString, save}
import org.clulab.scala.BufferedIterator
import org.clulab.scala.WrappedArray._
import org.clulab.scala.WrappedArrayBuffer._
import ForwardLayer._

import scala.collection.mutable.ArrayBuffer

class GreedyForwardLayer (parameters:ParameterCollection,
                          inputSize: Int,
                          isDual: Boolean,
                          t2i: Map[String, Int],
                          i2t: Array[String],
                          H: Parameter,
                          rootParam: Parameter,
                          distanceEmbeddingSize: Int,
                          distanceLookupParameters: Option[LookupParameter],
                          nonlinearity: Int,
                          dropoutProb: Float)
  extends ForwardLayer(parameters, inputSize, isDual, t2i, i2t, H, rootParam,
    distanceEmbeddingSize, distanceLookupParameters, nonlinearity, dropoutProb) {

  override def loss(finalStates: ExpressionVector, goldLabelStrings: IndexedSeq[Label]): Expression = {
    val goldLabels = Utils.toIds(goldLabelStrings.map(_.label), t2i)
    Utils.sentenceLossGreedy(finalStates, goldLabels)
  }

  override def saveX2i(printWriter: PrintWriter): Unit = {
    save(printWriter, TYPE_GREEDY, "inferenceType")
    save(printWriter, inputSize, "inputSize")
    save(printWriter, if (isDual) 1 else 0, "isDual")
    save(printWriter, distanceEmbeddingSize, "distanceEmbeddingSize")
    save(printWriter, nonlinearity, "nonlinearity")
    save(printWriter, t2i, "t2i")
    save(printWriter, dropoutProb, "dropoutProb")
  }

  override def toString: String = {
    s"GreedyForwardLayer($inDim, $outDim)"
  }

  override def inference(emissionScores: Array[Array[Float]]): IndexedSeq[String] = {
    val labelIds = Utils.greedyPredict(emissionScores)
    labelIds.map(i2t(_))
  }

  override def inferenceWithScores(emissionScores: Array[Array[Float]]): IndexedSeq[IndexedSeq[(String, Float)]] = {
    // For each index of the second Array, pair the value (score) at the index with i2t at the index and then sort
    // the pairs from highest to lowest by the score.
    emissionScores.map { scoresForPosition =>
      i2t.zip(scoresForPosition).sortBy(-_._2).toIndexedSeq
    }
  }
}

object GreedyForwardLayer {
  def load(parameters: ParameterCollection,
           x2iIterator: BufferedIterator[String]): GreedyForwardLayer = {
    //
    // load the x2i info
    //
    val byLineIntBuilder = new ByLineIntBuilder()
    val byLineFloatBuilder = new ByLineFloatBuilder()
    val byLineStringMapBuilder = new ByLineStringMapBuilder()

    val inputSize = byLineIntBuilder.build(x2iIterator, "inputSize")
    val isDualAsInt = byLineIntBuilder.build(x2iIterator, "isDual", DEFAULT_IS_DUAL)
    val isDual = isDualAsInt == 1
    val distanceEmbeddingSize = byLineIntBuilder.build(x2iIterator, "distanceEmbeddingSize", 0)
    val nonlinearity = byLineIntBuilder.build(x2iIterator, "nonlinearity", ForwardLayer.NONLIN_NONE)
    val t2i = byLineStringMapBuilder.build(x2iIterator, "t2i")
    val i2t = fromIndexToString(t2i)
    val dropoutProb = byLineFloatBuilder.build(x2iIterator, "dropoutProb", ForwardLayer.DEFAULT_DROPOUT_PROB)

    //
    // make the loadable parameters
    //
    val actualInputSize = if(isDual) 2 * inputSize + distanceEmbeddingSize else inputSize

    val H = parameters.addParameters(Dim(t2i.size, actualInputSize))
    val rootParam = parameters.addParameters(Dim(inputSize))

    val distanceLookupParameters =
      if(distanceEmbeddingSize > 0) Some(parameters.addLookupParameters(101, Dim(distanceEmbeddingSize)))
      else None

    new GreedyForwardLayer(parameters,
      inputSize, isDual, t2i, i2t, H, rootParam,
      distanceEmbeddingSize, distanceLookupParameters,
      nonlinearity, dropoutProb)
  }
}

File: main/src/main/python/pytorch/greedyForwardLayer.py
from pytorch.forwardLayer import *
from pytorch.utils import *
import numpy as np

class GreedyForwardLayer(ForwardLayer):
    def __init__(self, inputSize, isDual, t2i, i2t, actualInputSize, nonlinearity, dropoutProb, spans = None):
        super().__init__(inputSize, isDual, t2i, i2t, actualInputSize, nonlinearity, dropoutProb, spans)

    def loss(self, finalStates, goldLabelStrings):
        goldLabels = [self.t2i[gs] for gs in goldLabelStrings]
        return sentenceLossGreedy(finalStates, goldLabels)

    def saveX2i(self):
        x2i = dict()
        x2i["inferenceType"] = TYPE_GREEDY
        x2i["inputSize"] = self.inputSize
        x2i["isDual"] = 1 if self.isDual else 0
        x2i["span"] = spanToString(self.spans) if self.spans else ""
        x2i["nonlinearity"] = self.nonlinearity
        x2i["t2i"] = self.t2i

        return x2i

    def __str__(self):
        return f"GreedyForwardLayer({self.inDim}, {self.outDim})"

    def inference(self, emissionScores):
        emissionScores = emissionScoresToArrays(emissionScores)
        return [self.i2t[np.argmax(es)] for es in emissionScores]

    def inference2(self, emissionScores):
        return torch.argmax(emissionScores, dim=1)

    def inferenceWithScores(self, emissionScores):
        emissionScores = emissionScoresToArrays(emissionScores)
        return [sorted([(i, s) for i, s in enumerate(scoresForPosition)], key=lambda x: x[1]) for scoresForPosition in emissionScores]

    @classmethod
    def load(cls, x2i):
        inputSize = x2i["inputSize"]
        isDual = x2i.get("isDual", DEFAULT_IS_DUAL) == 1
        sapnValue = x2i.get("span", "")
        spans = None if sapnValue == "" else parseSpan(sapnValue, inputSize)
        nonlinearity = x2i.get("nonlinearity", NONLIN_NONE)
        t2i = x2i["t2i"]
        i2t = {i:t for t, i in t2i.items()}
        dropoutProb = x2i.get("dropoutProb", DEFAULT_DROPOUT_PROBABILITY)

        if spans:
            l = spanLength(spans)
            actualInputSize = 2*l if isDual else l
        else:
            actualInputSize = 2*inputSize if isDual else inputSize

        return cls(inputSize, isDual, t2i, i2t, actualInputSize, nonlinearity, dropoutProb, spans)
    